source/transactions/tests/run-command.json

```diff
@@ -0,0 +1,205 @@
+              },
+              "startTransaction": true,
+              "autocommit": false,
+              "writeConcern": null
+            },
+            "command_name": "count",
+            "database_name": "transaction-tests"
+          }
+        },
+        {
+          "command_started_event": {
+            "command": {
+              "commitTransaction": 1,
+              "lsid": "session0",
+              "txnNumber": {
+                "$numberLong": "1"
+              },
+              "startTransaction": null,
+              "autocommit": false,
+              "writeConcern": null
+            },
+            "command_name": "commitTransaction",
+            "database_name": "admin"
+          }
+        }
+      ]
+    },
+    {
+      "description": "run command with explicit primary read preference",
+      "operations": [
+        {
+          "name": "startTransaction",
+          "arguments": {
+            "session": "session0"
+          }
+        },
+        {
+          "name": "runCommand",
+          "arguments": {
+            "command": {
+              "count": "test"
+            },
+            "session": "session0",
+            "readPreference": {
+              "mode": "Primary"
+            }
+          },
+          "result": {
+            "n": 0
+          }
+        },
+        {
+          "name": "commitTransaction",
+          "arguments": {
+            "session": "session0"
+          }
+        }
+      ],
+      "expectations": [
+        {
+          "command_started_event": {
+            "command": {
+              "count": "test",
+              "readConcern": null,
+              "lsid": "session0",
+              "txnNumber": {
+                "$numberLong": "1"
+              },
+              "startTransaction": true,
+              "autocommit": false,
+              "writeConcern": null
+            },
+            "command_name": "count",
+            "database_name": "transaction-tests"
+          }
+        },
+        {
+          "command_started_event": {
+            "command": {
+              "commitTransaction": 1,
+              "lsid": "session0",
+              "txnNumber": {
+                "$numberLong": "1"
+              },
+              "startTransaction": null,
+              "autocommit": false,
+              "writeConcern": null
+            },
+            "command_name": "commitTransaction",
+            "database_name": "admin"
+          }
+        }
+      ]
+    },
+    {
+      "description": "run command with explicit secondary read preference",
+      "operations": [
+        {
+          "name": "startTransaction",
+          "arguments": {
+            "session": "session0"
+          }
+        },
+        {
+          "name": "runCommand",
+          "arguments": {
+            "command": {
+              "count": "test"
+            },
+            "session": "session0",
+            "readPreference": {
+              "mode": "Secondary"
```

Comment:
The spec says that IF a generic runCommand is called with an EXPLICIT read preference, that the driver should use the explicit read preference overriding the transaction read preference.

This command should therefore result in an error (because for now transactions require read preference Primary).

The relevant quote from the spec is:

> For drivers that allow read concern to be passed to a particular operation, this specification says, "If the user supplies an explicit readConcern via a method option, however, drivers MUST apply the readConcern, which will result in a server error." The same principle applies to write concern and read preference. The alternative is to silently ignore the user's read concern (or other options) if it is prohibited with transactions, but this would be a surprising and undetectable deviation from the user's intent.

Reply:
@ShaneHarvey I'll leave this one for you to ponder

Reply:
Yeah, this test should fail, not succeed. It should use read preference secondary and then throw "read preference in a transaction must be primary". Let's update this test to check that the command indeed fails.

Let's add a second test that puts read preference secondary in clientOptions, read preference primary in the startTransaction arguments and then executes runCommand with no readPreference. That runCommand should succeed.

Reply:
The spec fragment that I based this on is:

> The RunCommand method is considered a read operation and MUST use the transaction’s read preference.

I'll wait for Shane to weigh in since he already approved this.

Reply:
@rstam I think you are confusing readConcern and readPreference. According to the spec runcommand should always use the transaction's readPreference.

Reply:
Oh, you're right Robert, I'm the one who is confused. I didn't realize that section of the spec was added.

Reply:
Done